aws/s3.go

```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
+
+		targetBuckets := output.Buckets[i:j]
+
+		var bucketCh = make(chan *S3Bucket, len(targetBuckets))
+		var wg sync.WaitGroup
 
-		if len(bucketNameSubStr) > 0 {
-			if !strings.Contains(*bucket.Name, bucketNameSubStr) {
+		for _, bucket := range targetBuckets {
+			if len(bucketNameSubStr) > 0 && !strings.Contains(*bucket.Name, bucketNameSubStr) {
 				logging.Logger.Debugf("Skipping - Bucket %s - failed substring filter - %s", *bucket.Name, bucketNameSubStr)
 				continue
 			}
-		}
 
-		if !excludeAfter.After(*bucket.CreationDate) {
-			logging.Logger.Debugf("Skipping - Bucket %s - matched CreationDate filter", *bucket.Name)
-			continue
+			wg.Add(1)
+			go func(bucket *s3.Bucket) {
+				defer wg.Done()
+				if err := getBucketInfo(svc, bucket, regionClients, bucketCh); err != nil {
+					logging.Logger.Warnf("Skipping - Bucket %s - Failed to get info - %s", *bucket.Name, err)
+				}
+			}(bucket)
 		}
 
-		bucketRegion, err := getS3BucketRegion(svc, bucket.Name)
-		if err != nil {
-			logging.Logger.Warnf("Skipping - Bucket %s - Failed to get bucket location.", *bucket.Name)
-			continue
-		}
+		go func() {
+			wg.Wait()
+			close(bucketCh)
+		}()
 
-		// Create session for current bucket region as bucket location's session is required for getting bucket tags.
-		if _, ok := regionClients[bucketRegion]; !ok {
-			logging.Logger.Debugf("Creating session - Bucket %s - region %s", *bucket.Name, bucketRegion)
-			awsSession, err := session.NewSession(&awsgo.Config{
-				Region: awsgo.String(bucketRegion)},
-			)
-			if err != nil {
-				return bucketNamesPerRegion, errors.WithStackTrace(err)
+		for b := range bucketCh {
+			if !b.InTargetRegion {
+				logging.Logger.Debugf("Skipping - Bucket %s - region - %s - not in target regions", *b.Name, *b.Region)
+				continue
 			}
-			regionClients[bucketRegion] = s3.New(awsSession)
-		}
 
-		validTags, excludeReason := hasValidTags(regionClients[bucketRegion], bucket)
-		if !validTags {
-			logging.Logger.Debugf("Skipping - Bucket %s - %s", *bucket.Name, excludeReason)
-			continue
-		}
+			if !excludeAfter.After(*b.CreationDate) {
+				logging.Logger.Debugf("Skipping - Bucket %s - matched CreationDate filter", *b.Name)
+				continue
+			}
+
+			if !hasValidTags(b.Tags) {
+				logging.Logger.Debugf("Skipping - Bucket %s - matched tag filter", *b.Name)
+				continue
+			}
 
-		if _, ok := bucketNamesPerRegion[bucketRegion]; !ok {
-			bucketNamesPerRegion[bucketRegion] = []*string{}
+			if _, ok := bucketNamesPerRegion[*b.Region]; !ok {
+				bucketNamesPerRegion[*b.Region] = []*string{}
+			}
+
+			bucketNamesPerRegion[*b.Region] = append(bucketNamesPerRegion[*b.Region], b.Name)
 		}
 
-		bucketNamesPerRegion[bucketRegion] = append(bucketNamesPerRegion[bucketRegion], bucket.Name)
+		batchCount++
 	}
 
 	return bucketNamesPerRegion, nil
 }
 
+func getBucketInfo(svc *s3.S3, bucket *s3.Bucket, regionClients map[string]*s3.S3, bucketCh chan<- *S3Bucket) error {
+	var b S3Bucket
```

Comment:
Please minimize usage of single-variable names. E.g., `bucketData` instead of `b`.

Reply:
Done